network/pubsubchannel/pubsub.go

```diff
@@ -0,0 +1,157 @@
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+package pubsubchannel
+
+import (
+	"strconv"
+	"sync"
+
+	"github.com/ethersphere/swarm/log"
+)
+
+//PubSubChannel represents a pubsub system where subscriber can .Subscribe() and publishers can .Publish() or .Close().
+type PubSubChannel struct {
+	subscriptions []*Subscription
+	subsMutex     sync.RWMutex
+	nextId        int
+	quitC         chan struct{}
+}
+
+// Subscription is created in PubSubChannel using pubSub.Subscribe(). Subscribers can receive using .ReceiveChannel().
+// or .Unsubscribe()
+type Subscription struct {
+	closed    bool
+	pubSubC   *PubSubChannel
+	signal    chan interface{}
+	closeOnce sync.Once
+	id        string
+	lock      sync.RWMutex
+}
+
+// New creates a new PubSubChannel.
+func New() *PubSubChannel {
+	return &PubSubChannel{
+		subscriptions: make([]*Subscription, 0),
+		quitC:         make(chan struct{}, 1),
+	}
+}
+
+// Subscribe creates a subscription to a channel, each subscriber should keep its own Subscription instance.
+func (psc *PubSubChannel) Subscribe() *Subscription {
+	psc.subsMutex.Lock()
+	defer psc.subsMutex.Unlock()
+	newSubscription := newSubscription(strconv.Itoa(psc.nextId), psc)
+	psc.nextId++
+	psc.subscriptions = append(psc.subscriptions, &newSubscription)
+
+	return &newSubscription
+}
+
+func (psc *PubSubChannel) removeSub(s *Subscription) {
+	psc.subsMutex.Lock()
+	defer psc.subsMutex.Unlock()
+
+	for i, subscription := range psc.subscriptions {
+		if subscription.signal == s.signal {
+			log.Debug("Unsubscribing", "id", subscription.id)
+			subscription.lock.Lock()
+			subscription.closed = true
+			subscription.lock.Unlock()
+			psc.subscriptions = append(psc.subscriptions[:i], psc.subscriptions[i+1:]...)
+		}
+	}
+}
+
+// Publish broadcasts a message asynchronously to each subscriber.
+// If some of the subscriptions(channels) has been marked as closeable, it does it now.
+func (psc *PubSubChannel) Publish(msg interface{}) {
+	psc.subsMutex.RLock()
+	defer psc.subsMutex.RUnlock()
+	for _, sub := range psc.subscriptions {
+		sub.lock.RLock()
+		if sub.closed {
+			log.Debug("Subscription was closed", "id", sub.id)
+			sub.closeChannel()
+		} else {
+			go func(sub *Subscription) {
+				select {
+				case sub.signal <- msg:
+				case <-psc.quitC:
+				}
+			}(sub)
+		}
+		sub.lock.RUnlock()
+	}
+}
+
+// NumSubscriptions returns how many subscriptions are currently active.
+func (psc *PubSubChannel) NumSubscriptions() int {
+	psc.subsMutex.RLock()
+	defer psc.subsMutex.RUnlock()
+	return len(psc.subscriptions)
+}
+
+// Close cancels all subscriptions closing the channels associated with them.
+// Usually the publisher is in charge of calling Close().
+func (psc *PubSubChannel) Close() {
+	psc.subsMutex.Lock()
+	defer psc.subsMutex.Unlock()
+	for _, sub := range psc.subscriptions {
+		sub.lock.Lock()
+		sub.closed = true
+		sub.closeChannel()
+		sub.lock.Unlock()
+	}
+	psc.quitC <- struct{}{}
```

Comment:
This will terminate only one goroutine.

Also it will terminate one goroutine when PubSubChannel is closed, not when a subscription is unsubscribed. Is it possible that goroutines are left waiting even after unsubscription?

Reply:
Sorry, I don't know what I was thinking. I have changed the channel to size 0 and closing the channel when the PubSubChannel is closed.

Reply:
Ok, let me see if I can fix this

Reply:
Thank you.

Reply:
I have modified Unsubscribe to close the blocked publishing channels. 
The second use case, I'm not sure is too much relevant, I will try to fix it also.